Repo

luca020400 opened the pull request.

luca020400 committed:
Glimpse: GlimpseCursorLoader: Enforce non-null queryArgs

Change-Id: I35fa64c8bd03eddc98ed15e728cd8e43d06447c7

luca020400 committed:
Glimpse: MEDIA_STORE_REELS_LOADER_ID -> MEDIA_STORE_MEDIA_LOADER_ID

Change-Id: I4c626456b29f841aaa7b4b9cc5df03b6a1bd80c4

luca020400 committed:
Glimpse: Only specify the necessary projections to AlbumsProjection

Change-Id: I6289cc0f35bce8a28fb6cf8de3da725b11a819ab

luca020400 committed:
Glimpse: Implement data loading using the Repository/ViewModel pattern

Change-Id: Ia152000673b36e7f690ac5f2665ef1a0204bb339

luca020400 committed:
Glimpse: Add bucket ID to Media

Change-Id: I0f94f8444ac11a65a68e7913f29c3208a03c315c

luk1337 commented on app/src/main/java/androidx/loader/content/GlimpseCursorLoader.kt:
```diff
@@ -25,7 +25,7 @@ class GlimpseCursorLoader(
     selectionArgs: Array<out String>?,
     sortOrder: String?,
-    private val queryArgs: Bundle?
+    private val queryArgs: Bundle? = Bundle()
```
doesn't need to be nullable anymore if you want to do this.

luk1337 commented on app/src/main/java/org/lineageos/glimpse/fragments/AlbumFragment.kt:
```diff
@@ -64,13 +61,18 @@ class AlbumFragment : Fragment(R.layout.fragment_album), LoaderManager.LoaderCal
+                    mediaViewModel.setBucketId(album.id)
+                    mediaViewModel.mediaForAlbum.collect { data ->
+                        thumbnailAdapter.data = data.toTypedArray()
+
```
empty line

luk1337 commented on app/src/main/java/org/lineageos/glimpse/fragments/AlbumFragment.kt:
```diff
@@ -118,80 +120,14 @@ class AlbumFragment : Fragment(R.layout.fragment_album), LoaderManager.LoaderCal
+            viewLifecycleOwner.lifecycleScope.launch {
+                mediaViewModel.mediaForAlbum.collect { data ->
+                    thumbnailAdapter.data = data.toTypedArray()
 
```
empty line?

luk1337 commented on app/src/main/java/org/lineageos/glimpse/GlimpseApplication.kt:
```diff
@@ -13,8 +13,11 @@ import coil.decode.GifDecoder
+import org.lineageos.glimpse.repository.MediaRepository
 
 class GlimpseApplication : Application(), ImageLoaderFactory {
+    val mediaRepository: MediaRepository = MediaRepository(this)
```
no need to specify type?

luca020400 commented:
This is ready now.
It's based of the MediaDialog PR since it changes the data stored/fetched and I didn't want to re-do it.

luk1337 commented on app/src/main/java/org/lineageos/glimpse/viewmodels/MediaViewModel.kt:
```diff
@@ -0,0 +1,59 @@
+
+        val Factory: ViewModelProvider.Factory = viewModelFactory {
+            initializer {
+                val savedStateHandle = createSavedStateHandle()
```
inline them?

luk1337 commented on app/src/main/java/org/lineageos/glimpse/viewmodels/MediaViewModel.kt:
```diff
@@ -0,0 +1,59 @@
+class MediaViewModel(
+    private val savedStateHandle: SavedStateHandle, private val mediaRepository: MediaRepository
+) : ViewModel() {
+    private val _mediaPosition = savedStateHandle.getLiveData<Int>(MEDIA_POSITION_KEY)
```
maybe "Internal" at the end instead of `_`

luk1337 commented on app/src/main/java/org/lineageos/glimpse/viewmodels/MediaViewModel.kt:
```diff
@@ -0,0 +1,59 @@
+    private val savedStateHandle: SavedStateHandle, private val mediaRepository: MediaRepository
+) : ViewModel() {
+    private val _mediaPosition = savedStateHandle.getLiveData<Int>(MEDIA_POSITION_KEY)
+    val mediaPositionLiveData: LiveData<Int> = _mediaPosition
```
types not needed

luca020400 replied:
this is "needed", mediaPositionInternal is MutableLiveData, I want to expose a non mutable variant

luk1337 commented on app/src/main/java/org/lineageos/glimpse/thumbnail/ThumbnailAdapter.kt:
```diff
@@ -70,16 +72,12 @@ class ThumbnailAdapter(
-                getMediaFromMediaStore(truePosition)?.let {
-                    thumbnailViewHolder.bind(it, truePosition)
-                }
+                    thumbnailViewHolder.bind(data[truePosition], truePosition)
```
too many spaces

luk1337 commented on app/src/main/java/org/lineageos/glimpse/thumbnail/ThumbnailAdapter.kt:
```diff
@@ -25,35 +23,39 @@ import java.util.Date
 
-    override fun getItemId(position: Int) = getIdFromMediaStore(position)
+    override fun getItemId(position: Int) =
+        if (headersPositions.contains(position)) position.toLong() else data[getTruePosition(
```
would prefer to see {} here

luk1337 commented on app/src/main/java/org/lineageos/glimpse/repository/MediaFlow.kt:
```diff
@@ -0,0 +1,126 @@
+                        )
+                        it.moveToNext()
+                    }
+                    send(media)
```
declare media inside data?.let

luca020400 replied:
actually no, I have to move it outside of the let, I need to send a zero lenght list in case there's no items in the cursor

luk1337 commented on app/src/main/java/org/lineageos/glimpse/repository/MediaFlow.kt:
```diff
@@ -0,0 +1,126 @@
+    fun flow(): Flow<List<Media>> = callbackFlow {
+        val projection = MediaQuery.MediaProjection
+        val imageOrVideo =
+            (MediaStore.Files.FileColumns.MEDIA_TYPE eq MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE) or (MediaStore.Files.FileColumns.MEDIA_TYPE eq MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO)
```
too long

luk1337 commented on app/src/main/java/org/lineageos/glimpse/repository/AlbumsFlow.kt:
```diff
@@ -0,0 +1,119 @@
+        val onLoadCompleteListener = OnLoadCompleteListener<Cursor> { _, data: Cursor? ->
+            if (!isActive) return@OnLoadCompleteListener
+            launch(Dispatchers.IO) {
+                val albums = mutableMapOf<Int, Album>()
```
declare inside data?.let

luca020400 replied:
same as MediaFlow

luk1337 commented on app/src/main/java/org/lineageos/glimpse/models/Album.kt:
```diff
@@ -5,10 +5,11 @@
 import android.net.Uri
 import android.os.Parcel
 import android.os.Parcelable
-import org.lineageos.glimpse.ext.*
```
keep `*`

luk1337 commented on app/src/main/java/org/lineageos/glimpse/fragments/ReelsFragment.kt:
```diff
@@ -59,19 +58,21 @@ class ReelsFragment : Fragment(R.layout.fragment_reels), LoaderManager.LoaderCal
             parentNavController.navigate(
-                R.id.action_mainFragment_to_mediaViewerFragment,
-                MediaViewerFragment.createBundle(
+                R.id.action_mainFragment_to_mediaViewerFragment, MediaViewerFragment.createBundle(
```
unneessary code style change?

luk1337 commented on app/src/main/java/org/lineageos/glimpse/fragments/MediaViewerFragment.kt:
```diff
@@ -84,7 +80,15 @@ class MediaViewerFragment : Fragment(
+                        mediaViewerAdapter.data = data.toTypedArray()
+                        viewPager.setCurrentItem(mediaViewModel.mediaPosition, false)
+                        onPageChangeCallback.onPageSelected(mediaViewModel.mediaPosition)
+
```
empty line

luk1337 commented on app/src/main/java/org/lineageos/glimpse/fragments/AlbumsFragment.kt:
```diff
@@ -77,106 +72,12 @@ class AlbumsFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {
+        viewLifecycleOwner.lifecycleScope.launch {
+            mediaViewModel.albums.collect() {
+                albumThumbnailAdapter.changeArray(it.toTypedArray())
 
```
empty line

luca020400 commented on app/src/main/java/org/lineageos/glimpse/fragments/AlbumFragment.kt:
```diff
@@ -118,80 +119,13 @@ class AlbumFragment : Fragment(R.layout.fragment_album), LoaderManager.LoaderCal
-        }
-    }
-
-    override fun onConfigurationChanged(newConfig: Configuration) {
```
I removed a bit too much, restore

luca020400 commented on app/src/main/java/org/lineageos/glimpse/fragments/MediaViewerFragment.kt:
```diff
@@ -96,19 +95,10 @@ class MediaViewerFragment : Fragment(
-        set(value) {
-            currentPositionLiveData.value = value
-        }
     private val album by lazy { arguments?.getParcelable(KEY_ALBUM, Album::class) }
```
We may want to move that to viewModel too like position and use that internally to pick the proper bucketId

luca020400 commented on app/src/main/java/org/lineageos/glimpse/viewmodels/MediaViewModel.kt:
```diff
@@ -0,0 +1,51 @@
+class MediaViewModel(
+    private val savedStateHandle: SavedStateHandle, private val mediaRepository: MediaRepository
+) : ViewModel() {
+    private val mediaPositionInternal = savedStateHandle.getLiveData<Int>(MEDIA_POSITION_KEY)
```
I'd love to use flows here as well, we can use getInt and get the raw data and set it to a flow, should work :)

luca020400 commented on app/src/main/java/org/lineageos/glimpse/repository/MediaFlow.kt:
```diff
@@ -0,0 +1,129 @@
+import org.lineageos.glimpse.utils.MediaStoreRequests
+
+class MediaFlow(private val context: Context, private val bucketId: Int?) {
+    fun flow() = callbackFlow {
```
no, add back the type, callbackFlow by default is a mutableFlow, we want a Flow
ditto for albums

SebaUbuntu replied:
![image](https://github.com/SebaUbuntu/android_packages_apps_Glimpse/assets/38215111/5994f6f4-fbde-4757-aa85-27698f5a717b)

luca020400 replied:
nvm then, maybe it was something I had early on

luca020400 committed:
Glimpse: Don't use CursorLoader

Instead manually invoke the query and listen for updates
on the external media Uri to trigger updates.

Change-Id: I0d86484dfa453cce9910d2bcadee6191de5aed28

luca020400 commented:
@SebaUbuntu last commit should be pushed as is, but I'm too deep to go back to upstream atm

luca020400 commented:
@SebaUbuntu I think this is good to go, paging doesn't seem an easy feat to implement without a lot of thinking/work
I tried up to 25k items, and it's perfect

luca020400 committed:
Glimpse: Switch AlbumThumbnailAdapter to `var data`

A bit nicer instead of nullable list

Change-Id: If56620b5883c0e3b53a7fa0d62311c2e0999ca66

luca020400 committed:
Glimpse: Conflate any flow from URIs

If we get a lot of new files we want to drop any "old" event

Change-Id: I913672d4b651ca0473edc530e0f57996fd4b6830

luca020400 committed:
Glimpse: Collect on resume

ContentObservers are tied to the lifecycle of the application
and must be started/stopped onResume to ensure proper callbacks

Change-Id: I63c1a6ebc3e734a84aa729f36296d570801e6c70
